### ISSUE TITLE
Register-PSSessionConfiguration requires writing to $PSHOME\SessionConfig folder

### ISSUE THREAD

**Issue by SteveL-MSFT** [PowerShell/PowerShell#9278]
# Summary of the new feature/enhancement

`Register-PSSessionConfiguration` stores the config file in [$PSHOME\SessionConfig](https://github.com/PowerShell/PowerShell/search?l=C%23&q=%22sessionconfig%22) folder.

If PSCore6 is an AppX, even an elevated admin won't have write access to $PSHOME.  This means that `Register-PSSessionConfiguration` can't work.  Should find another location that only admins can write to that WinRM can read from for this cmdlet.

**Comment by iSazonov** [comment]
Long time Windows exposes UAC virtualization for registry keys and file system. 
If we could use it it would solve this problem.

**Comment by SteveL-MSFT** [comment]
@iSazonov that would potentially work for Windows, but doesn't solve the same problem with Snaps.

**Comment by iSazonov** [comment]
I like the way for Windows. I think it is right way. 
I guess there is a solution in Snaps too - PowerShell is not single application with such behavior. My quick search https://forum.snapcraft.io/t/snap-layouts/7207 "Creating new files and directories in read-only spaces". So I believe Snaps has that we need.

**Comment by TravisEz13** [comment]
What parts of snap would we map?  Just because we can, doesn't mean we should.  We should not make any part of $PSHome writable on linux.

**Comment by bcdady** [comment]
See also https://github.com/PowerShell/PowerShell-Snap/issues/32

**Comment by joeyaiello** [comment]
Can we just use `/etc/` for this? Or does Snap "classic" et al not have a way to do this?

**Comment by TravisEz13** [comment]
@joeyaiello The problem is this should be done in PowerShell. `/etc` is where the file should be for all PowerShell installations.

**Comment by iSazonov** [comment]
For reference #4080

**Comment by TravisEz13** [comment]
Known locations that write into $PSHOME - In-progress

| Item                                                      | Location|
|------------------------------------| ----------|
|`Register-PSSessionConfiguration`|`$PSHOME\SessionConfig` |
|`Update-Help -scope currentuser` | [May just be a bug](https://github.com/PowerShell/PowerShell-Snap/issues/32) |
|`Update-Help -scope AllUsers` | ? |
|`Enable-ExperimentalFeature -Scope AllUsers`| `$pshome\powershell.config.json` |
|`Set-ExecutionPolicy -Scope LocalMachine`| `$pshome\powershell.config.json` |

**Comment by TravisEz13** [comment]
## Proposal

### Rational

It looks like the majority of the problem is writing to the configuration file.  We should move it under `/etc` for Linux and macOS, and `%programdata% for Windows.

We can make similar changes for `$PSHome\SessionConfig` and `$PSHome\en-US` (and related folders) later.

### Details

#### Security

These folders should be secured similarly to the currently location the file is located at when the product is installed via DEB, RPM, or MSI.

The configuration file is system wide configuration and only system administrators should be able to modify it.  This also eliminates pwsh creating it on first launch.

#### Linux and macOS

We would move the system configuration to `/etc/powershell/powershell.config.json`
We would use the folder so that any addition items will be group together.

Packages would have to be updated to create this and the zip instruction would have to be update or a script created.

#### Windows

We would move the system configuration to `%programdata%\Microsoft\PowerShell\powershell.config.json`

MSI could be update to create the file.
MSIX - @stevel-msft?
zip - instructions would have to be updated

### Issues

1. What do you do on upgrade?  Currently we overwrite the file, but this is not the correct behavior.
2. Not sure how to set the default config after MSIX install.
   * Possible solution is to include a script that admins can run after install.  This would simplify zip and tar.gz install as well.

**Comment by SteveL-MSFT** [comment]
For PowerShell.config.json, it might be better to have cmdlet(s) to manage that file with `-Scope AllUsers/CurrentUser`.  This would work cross-platform.

**Comment by iSazonov** [comment]
Cmdlets is an option. The presence of configs implies supporting direct editing.

**Comment by TravisEz13** [comment]
@SteveL-MSFT  I don't understand your point at all.  Most of those already exist.  The problem is that the AllUser file needs to be writable.

**Comment by SteveL-MSFT** [comment]
@TravisEz13 system wide config can exist, but needs to be under ProgramData (on Windows) and requires elevation to set.  My point of the cmdlets isn't for setting individual settings, but something like `Get-PowerShellConfig` which returns an object representing the config and `Set-PowerShellConfig` to save it, but with `-Scope` parameter so the user doesn't need to know where in the filesystem it goes.

**Comment by daxian-dbw** [comment]
> We would move the system configuration to /etc/powershell/powershell.config.json
We would move the system configuration to %programdata%\Microsoft\PowerShell\powershell.config.json

This means different versions of `pwsh` will share the same configuration file, which is problematic. Even with sub-folder named after the version, it means different instances of the same version will share the same configuration file, which is still a breaking change from today's behavior.

**Comment by TravisEz13** [comment]
@SteveL-MSFT I still don't see how this addresses this issue at all.  I think you have a fundamental mis-understanding of the problem.  That might be a useful new feature, but it doesn't help solve this problem.

**Comment by TravisEz13** [comment]
@daxian-dbw System wide configuration should be actually system wide, this is a reason to move it into a common folder.  I can see reason to have different files for preview and stable.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this, we agree not to take any changes for RC.  We should relnote the limitations and address post 7.0.  The proposal moving the system level config to /etc and %programdata% makes sense.

**Comment by iSazonov** [comment]
In last days I make some experiments with the config and new Core Json API (to remove dependency from Newtonsoft Json.NET) and discover some inconsistencies in the config file format and serialization/deserialization. I hope we fix this too in 7.1.

**Comment by joeyaiello** [comment]
I think we need to make this a priority for 7.1, but particularly implementing the `/etc/powershell/powershell.config.json` . The scenario is essentially that, as on Windows with the GP store, Unix admins need to be confident that Computer-wide Policy is applied to all PowerShells on a box, irrespective of whether they were bin-dropped as portable copies or installed to XDG-compliant locations via a package.

I can follow up with offline justification if needed

**Comment by iSazonov** [comment]
Now .Net Core 5.0 has new enhancements in JSON implementation and this raises a question - will we move on .Net Core 5.0 in 7.1 timeframe and benefit from new .Net Core enhancements and improvements? 
PowerShell Core 7.0 LTS is based on .Net Core 3.1 LTS - release in 2020 year
PowerShell Core 7.1 GA is based on .Net Core 5.0 GA - release in 2021 year
PowerShell Core 8.0 LTS is based on .Net Core 6.0 LTS - release in 2022 year

**Comment by SteveL-MSFT** [comment]
@iSazonov the plan is to move to .NET 5 for 7.1 as soon as they have a public preview.

**Comment by doctordns** [comment]
@SteveL-MSFT - would you release a 7.1 on a pre-release version of .NET core 5?

**Comment by iSazonov** [comment]
@doctordns Release 7.0 will be on release .Net Core 5.

**Comment by TravisEz13** [comment]
https://github.com/PowerShell/PowerShell/issues/9278#issuecomment-487822446
The suggested workaround here does not work with snaps in classic containment (full access to the host)

**Comment by BouwenMA** [comment]
Not seeing much reference to also being unable to create a global profile.ps1 due to $PSHome being unwritable for Windows.  I found issue #[12519](https://github.com/PowerShell/PowerShell/issues/12519) "Unable to create profile.ps1" that was opened about this, and later closed referencing this issue, but never had feedback noted here...  So adding that here now myself as I've also come across the same problems as well..

**Comment by aderusha** [comment]
This issue has been open now for 4 years.  When you go to [Microsoft's page for installing Powershell](https://learn.microsoft.com/en-us/powershell/scripting/install/installing-powershell-on-windows?view=powershell-7.3), the very first section is entitled: **Install PowerShell using Winget (recommended)**

If you do so, the installation is broken because nobody can adjust things like the execution policy, including administrators.

Is this ever going to be fixed?  MS is directly telling everybody to deploy PS using an install which doesn't work and hasn't worked for years.

**Comment by erelado** [comment]
> This issue has been open now for 4 years. When you go to [Microsoft's page for installing Powershell](https://learn.microsoft.com/en-us/powershell/scripting/install/installing-powershell-on-windows?view=powershell-7.3), the very first section is entitled: **Install PowerShell using Winget (recommended)**
> 
> If you do so, the installation is broken because nobody can adjust things like the execution policy, including administrators.
> 
> Is this ever going to be fixed? MS is directly telling everybody to deploy PS using an install which doesn't work and hasn't worked for years.

Is there any news on this issue? What is the reason for the difficulty in fixing it already?

**Comment by ChrisGuzak** [comment]
+1.

**Comment by mrjester888** [comment]
Just ran into this issue.  Incredibly annoying to discover this has been a known problem for 4 years and MSFT hasn't bothered to fix it.

**Comment by jurjen74** [comment]
Just ran into this issue. ... reinstalled from MSI and now problems.

**Comment by djmevans** [comment]
Just encountered this as well on a new build and installed via the msix package instead of the msi. Windows Store version still prevents changing the execution policy.

**Comment by lafe** [comment]
I encountered this issue with the `Set-ExecutionPolicy` Cmdlet just this week. So from my point of view, it's still relevant.

**Comment by ssg** [comment]
I'm having the problem of not being able to `Set-ExecutionPolicy` because it tries to modify `powershell.config.json` in the Appx installation directory. This happens on an Administrator Powershell prompt:

```
>Set-ExecutionPolicy RemoteSigned

Set-ExecutionPolicy: Access to the path 'C:\Program Files\WindowsApps\Microsoft.PowerShell_7.6.0.0_x64__8wekyb3d8bbwe\powershell.config.json' is denied.
To change the execution policy for the default (LocalMachine) scope, start PowerShell with the "Run as administrator" option. To change the execution policy for the current user, run "Set-ExecutionPolicy -Scope CurrentUser".
```

**Comment by ssg** [comment]
@SteveL-MSFT marked the original bug as duplicate, now we're left at the mercy of triage for this issue marked as "Issue-Enhancement". It feels like an obvious bug to me at the moment, regression even.

This keeps happening on pwsh 7.6.0 installed from MSStore.

**Comment by SteveL-MSFT** [comment]
https://learn.microsoft.com/en-us/uwp/schemas/appxpackage/uapmanifestschema/element-desktop8-mutablepackagedirectory might be a way to address this

<blockquote><img src="https://learn.microsoft.com/en-us/media/open-graph-image.png" width="48" align="right"><div><strong><a href="https://learn.microsoft.com/en-us/uwp/schemas/appxpackage/uapmanifestschema/element-desktop8-mutablepackagedirectory">desktop8:MutablePackageDirectory - Windows UWP applications</a></strong></div><div>Enables your desktop application to specify a folder where you can modify the installation files for your application.</div></blockquote>

**Comment by DrusTheAxe** [comment]
> On Windows 10, it would make sense to be in $env:SystemDrive\ProgramData

Better answer is ApplicationData's MachineFolder

https://github.com/microsoft/WindowsAppSDK/blob/main/specs/applicationdata/ApplicationData.md#342-manifested-opt-in

TL;DR in `AppxManifest.xml` declare "I want a per-machine (machine-wide-accessible) data location".

1. Declare `<appdata:ApplicationData><MachineFolder/>` in `AppxManifest.xml`. Add the [Sddl=... attribute](https://github.com/microsoft/WindowsAppSDK/blob/main/specs/applicationdata/ApplicationData.md#341-machine-pathfolder) needed to grant additional permissions
2. Use [Microsoft.Windows.Storage.ApplicationData.MachinePath](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.windows.storage.applicationdata.machinepath?view=windows-app-sdk-1.8) to get its path

ApplicationData provides per-packagefamily per-user data storage.

ApplicationData.MachineFolder provides per-packagefamily **per-machine** data storage.

<blockquote><img src="https://repository-images.githubusercontent.com/256049233/d3399d67-2630-4f8c-8c3e-c6bcf3012725" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/microsoft/WindowsAppSDK/blob/main/specs/applicationdata/ApplicationData.md">WindowsAppSDK/specs/applicationdata/ApplicationData.md at main · microsoft/WindowsAppSDK</a></strong></div><div>The Windows App SDK empowers all Windows desktop apps with modern Windows UI, APIs, and platform features, including back-compat support, shipped via NuGet. - microsoft/WindowsAppSDK</div></blockquote>

**Comment by DrusTheAxe** [comment]
> I'm having the problem of not being able to `Set-ExecutionPolicy` because it tries to modify `powershell.config.json` in the Appx installation directory. This happens on an Administrator Powershell prompt:

By design. MSIX package's install location is not Writeable; Read (and eXecute) rights only. Else it would break MSIX's security and integrity promises, and preclude data efficiencies (e.g. shared by all users, not a unique copy/disk-space per user).

Some unpackaged apps use the classic pattern of "store config file next to app and modify the config file as/when desired". Powershell is one such example.

MSIX recommends storing an original/unmodified/template file in the package and storing an _override_ file in a mutable location e.g. [ApplicationData.LocalPath](https://learn.microsoft.com/windows/windows-app-sdk/api/winrt/microsoft.windows.storage.applicationdata.localpath) for per-user storage, or [ApplicationData.MachinePath](https://learn.microsoft.com/windows/windows-app-sdk/api/winrt/microsoft.windows.storage.applicationdata.machinepath) for per-machine storage. Then at runtime instead of the classic

```
f = OpenFile(path_to_exe + "\foo.data")
```

use the override pattern e.g.

```
f = OpenFile(ApplicationData.Current.LocalPath + "\foo.data")
IF not found
    f = OpenFile(path_to_exe + "foo.data")
```

**Comment by ssg** [comment]
> MSIX recommends storing an original/unmodified/template file in the package and storing an _override_ file in a mutable location e.g.

So, is it a bug with `Set-ExecutionPolicy`?

**Comment by DrusTheAxe** [comment]
> > MSIX recommends storing an original/unmodified/template file in the package and storing an _override_ file in a mutable location e.g.
> 
> So, is it a bug with `Set-ExecutionPolicy`?

That would be a fair characterization.

The rules for packaged and unpackaged processes aren't identical. This is one such example.

One _can_ make an application run in both contexts e.g. use override file [if running with package identity](https://devblogs.microsoft.com/insidemsix/is-this-a-packaged-process/). It's up the the app/dev to decide if this is warranted for their needs.
